examples/src/main/java/kafka/examples/ExactlyOnceMessageProcessor.java

```diff
@@ -79,111 +86,114 @@ public ExactlyOnceMessageProcessor(final String inputTopic,
+            int transactionTimeoutMs = 10_000;
+            KafkaProducer<Integer, String> producer =
+                new Producer(outputTopic, true, transactionalId, true, -1, transactionTimeoutMs, null).get();
+
+            // consumer must be in read_committed mode, which means it won't be able to read uncommitted data
+            boolean readCommitted = true;
+            KafkaConsumer<Integer, String> consumer = new Consumer(
+                "processor-consumer", bootstrapServers, inputTopic, "processor-group", Optional.of(groupInstanceId), readCommitted, -1, null)
+                    .createKafkaConsumer();
+
+            // called first and once to fence zombies and abort any pending transaction
+            producer.initTransactions();
+
+            consumer.subscribe(singleton(inputTopic), this);
+
+            Utils.printOut("Processing new records");
+            while (!closed && remainingRecords > 0) {
+                try {
+                    ConsumerRecords<Integer, String> records = consumer.poll(ofMillis(200));
+                    if (!records.isEmpty()) {
+                        // begin a new transaction session
+                        producer.beginTransaction();
+
+                        for (ConsumerRecord<Integer, String> record : records) {
+                            // process the record and send downstream
+                            ProducerRecord<Integer, String> newRecord =
+                                new ProducerRecord<>(outputTopic, record.key(), record.value() + "-ok");
+                            producer.send(newRecord);
+                        }
+
+                        // checkpoint the progress by sending offsets to group coordinator broker
+                        // note that this API is only available for broker >= 2.5
+                        producer.sendOffsetsToTransaction(getOffsetsToCommit(consumer), consumer.groupMetadata());
+
+                        // commit the transaction including offsets
+                        producer.commitTransaction();
+                        processedRecords += records.count();
                     }
-
-                    Map<TopicPartition, OffsetAndMetadata> offsets = consumerOffsets();
-
-                    // Checkpoint the progress by sending offsets to group coordinator broker.
-                    // Note that this API is only available for broker >= 2.5.
-                    producer.sendOffsetsToTransaction(offsets, consumer.groupMetadata());
-
-                    // Finish the transaction. All sent records should be visible for consumption now.
-                    producer.commitTransaction();
-                    messageProcessed += records.count();
+                } catch (AuthorizationException | UnsupportedVersionException | ProducerFencedException
+                         | FencedInstanceIdException | OutOfOrderSequenceException | SerializationException e) {
```

Comment:
Might need to be consistent with the producer change (i.e. check instanceOf retriableException)

Reply:
Well, this is consistent with the consumer. The exception list here is shorter and much more compact (no repeated instanceof). If you don't mind, I would keep it as it is.

Reply:
Good to me.